src/cpp/session/modules/SessionPackages.R

```diff
@@ -1226,79 +772,692 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
-   if (length(appendNewLineIndex))
-      Rproj[appendNewLineIndex] <- "AutoAppendNewline: Yes"
-   else
-      Rproj <- c(Rproj, "AutoAppendNewline: Yes")
+   # For packages that were installed from CRAN or RSPM, browse
+   # to the package page associated with the repository.
+   repository <- pkgDesc[["Repository"]]
+   if (identical(repository, "RSPM") || identical(repository, "CRAN"))
+   {
+      repoUrl <- .rs.nullCoalesce(
+         pkgDesc[["RemoteRepos"]],
+         repos[[repository]],
+         repos[[1L]]
+      )
+      
+      if (length(repoUrl))
+      {
+         url <- sprintf("%s/package=%s", repoUrl[[1L]], pkgDesc[["Package"]])
+         return(url)
+      }
+   }
    
-   stripTrailingWhitespace <- grep("StripTrailingWhitespace:", Rproj, fixed = TRUE)
-   if (length(stripTrailingWhitespace))
-      Rproj[stripTrailingWhitespace] <- "StripTrailingWhitespace: Yes"
-   else
-      Rproj <- c(Rproj, "StripTrailingWhitespace: Yes")
+   # Try to build a link for packages installed from other remotes.
+   host <- pkgDesc[["RemoteHost"]]
+   user <- pkgDesc[["RemoteUsername"]]
+   repo <- pkgDesc[["RemoteRepo"]]
    
-   cat(Rproj, file = RprojPath, sep = "\n")
+   if (!is.null(host) && !is.null(user) && !is.null(repo))
+   {
+      host <- sub("api.github.com", "github.com", host, fixed = TRUE)
+      url <- sprintf("https://%s/%s/%s", host, user, repo)
+      return(url)
+   }
    
-   # compute scratch paths
-   scratchPaths <- .Call("rs_computeScratchPaths", RprojPath, PACKAGE = "(embedding)")
-   scratchPath <- scratchPaths$scratch_path
+   # If 'Repository' appears to be a URL, use it directly.
+   if (is.character(repository) && grepl("^https://", repository))
+   {
+      return(repository)
+   }
    
-   # NOTE: this file is not always generated (e.g. people who have implicitly opted
-   # into using devtools won't need the template file)
-   if (!is.null(scratchPath) &&
-       file.exists(file.path(packageDirectory, "R", "hello.R")))
+   # If this appears to be a Bioconductor package, link there.
+   if (!is.null(pkgDesc[["biocViews"]]))
    {
-      .Call("rs_addFirstRunDoc", scratchPath, "R/hello.R", PACKAGE = "(embedding)")
+      mirror <- getOption("BioC_mirror", default = "https://bioconductor.org")
+      fmt <- "%s/packages/release/bioc/html/%s.html"
+      url <- sprintf(fmt, mirror, pkgDesc[["Package"]])
+      return(url)
    }
+   
+   # Give up.
+   ""
+})
 
-   ## NOTE: This must come last to ensure the other package
-   ## infrastructure bits have been generated; otherwise
-   ## compileAttributes can fail
-   if (usingRcpp &&
-       .rs.isPackageVersionInstalled("Rcpp", "0.10.1") &&
-       require(Rcpp, quietly = TRUE))
-   {
-      Rcpp::compileAttributes(packageDirectory)
-      if (!is.null(scratchPath) &&
-          file.exists(file.path(packageDirectory, "src/rcpp_hello.cpp")))
+.rs.addFunction("listInstalledPackages", function()
```

Comment:
The prior implementation was re-written here to avoid some awkwardness in how package information was built and later restructured.